### PR TITLE
Fix cuda bootstrap for N = 8192

### PR DESCRIPTION
**Pull request by agnesLeroy**
### Checklist 

(Use '[x]' to check the checkboxes, or submit the PR and then click the checkboxes)

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] The PR description links to the related issue (to link an issue, use '#XXX'.)
* [ ] The tests on AWS have been launched and are successful (comment with @slab-ci cpu_test and/or @slab-ci gpu_test to trigger the tests)
* [ ] The draft release description has been updated
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by agnesLeroy** [comment]
@slab-ci cpu_test
@slab-ci gpu_test

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

**Review by bbarbakadze — CHANGES_REQUESTED** [review]

File: concrete-cuda/cuda/src/crypto/bootstrapping_key.cuh [hunk 1]
```diff
@@ -114,30 +115,75 @@ void cuda_convert_lwe_bootstrap_key(double2 *dest, ST *src, void *v_stream,
...
+      batch_NSMFFT<FFTDegree<Degree<8192>, ForwardFFT>, NOSM>
+          <<<gridSize, blockSize, 0, *stream>>>(d_bsk, dest, buffer);
+    }
     break;
   default:
     break;
   }
 
-  cudaFree(d_bsk);
+  cuda_drop_async(d_bsk, *stream, gpu_index);
```
[the remark this answers is not in the document]
Comment (agnesLeroy) [reply concrete-cuda/cuda/src/crypto/bootstrapping_key.cuh:186]: Yes definitely!

File: concrete-cuda/cuda/src/fft/bnsmfft.cuh [hunk 2]
```diff
@@ -718,10 +718,14 @@ __device__ void correction_inverse_fft_inplace(double2 *x) {
...
  * function takes as input already compressed input
  */
-template <class params>
-__global__ void batch_NSMFFT(double2 *d_input, double2 *d_output) {
-  extern __shared__ double2 sharedMemoryFFT[];
-  double2 *fft = sharedMemoryFFT;
+template <class params, sharedMemDegree SMD>
+__global__ void batch_NSMFFT(double2 *d_input, double2 *d_output,
+                             double2 *buffer) {
+  double2 *fft = buffer;
```
[the remark this answers is not in the document]
Comment (agnesLeroy) [reply concrete-cuda/cuda/src/fft/bnsmfft.cuh:724]: Indeed, this is replaced by
double2 *fft = &buffer[blockIdx.x * polynomial_size];

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

File: concrete-cuda/cuda/src/crypto/bootstrapping_key.cuh [hunk 3]
```diff
@@ -114,30 +115,75 @@ void cuda_convert_lwe_bootstrap_key(double2 *dest, ST *src, void *v_stream,
...
-    batch_NSMFFT<FFTDegree<Degree<512>, ForwardFFT>>
-        <<<gridSize, blockSize, shared_memory_size, *stream>>>(d_bsk, dest);
+    if (shared_memory_size <= cuda_get_max_shared_memory(gpu_index)) {
+      double2 *buffer = (double2 *)cuda_malloc_async(0, *stream, gpu_index);
+      batch_NSMFFT<FFTDegree<Degree<512>, ForwardFFT>, FULLSM>
+          <<<gridSize, blockSize, shared_memory_size, *stream>>>(d_bsk, dest,
+                                                                 buffer);
+    } else {
+      double2 *buffer =
+          (double2 *)cuda_malloc_async(shared_memory_size, *stream, gpu_index);
```
[the remark this answers is not in the document]
Comment (agnesLeroy) [reply concrete-cuda/cuda/src/crypto/bootstrapping_key.cuh]: definitely!

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

**Pull request merged**